### DISCUSSION TITLE
Problem while creating document with multi relationship field and file upload

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [pocketbase/pocketbase#2763] [Q&A] [answered]
Let say I have two collection plans and members. Members can have more than one plans as relation field with multiple option as well as there is picture field with file upload option. In flutter SDK, while trying to create document with picture plans field is sending as empty array while without picture working properly. Now as a work around two request is send to server one to create document and then update picture. Whether its known issue?

**Comment by ganigeorgiev** [comment]
Could you provide a code sample how you are sending the request?

Note that files must be added separately using the `files` named argument - https://github.com/pocketbase/dart-sdk#file-upload.

**Reply by iamdineshkumar** [reply]
Please find sample collection fields and dart code for creating document

Collection Details:
1 plans
- planName(varchar)
- validDays(number)
2. members
- memberName(varchar)
- plans(relations field with multiple option)
- picture(file for images)

Sample Dart Code
```dart
const data = {
    "memberName": "test",
    "plans": [
        "q2th2rhjeyn85od","vrnd4li9cnt9m9o"
    ],
};

const record = await pb.collection('members').create(body:data,
    files: [
        http.MultipartFile.fromString(
            'picture',
            'imagecontent.....',
            filename: 'test.png',
        ),
    ],);
```

**Reply by ganigeorgiev** [reply]
Hm, it seems that the multipart array data serialization is incorrect in https://github.com/pocketbase/dart-sdk/blob/master/lib/src/client.dart#L251 is incorrect. I'll try to investigate it in more details sometime later today.

**Reply by iamdineshkumar** [reply]
I tried those segment code in dart pad but it seems working fine sample code i used in dart pad
```dart

void main() {
  Map<String, dynamic> body ={'memberName': "demo Member", 
                               'plans': ['vc02zq08iftceag', 'dhrxlue2dhhlvlb']};
  Map<String, dynamic> fields = {};
  
    body.forEach((key, value) {
      if (value is Iterable) {
        for (var i = 0; i < value.length; i++) {
          fields["$key[$i]"] = value.elementAt(i).toString();
        }
      } else {
        fields[key] = value.toString();
      }
    });
  
    print(fields);
}

```

**Reply by ganigeorgiev** [reply]
The problem is not in the code itself, but how the field key is constructed when an array value is used.

When adding array values with multipart/form-data, array keys usually should be defined by adding the same key multiple times, eg.:

```
// key: ["a", "b", "c"]
formData.append("key", "a")
formData.append("key", "b")
formData.append("key", "c")
```

But [http.MultipartRequest](https://github.com/dart-lang/http/blob/master/pkgs/http/lib/src/multipart_request.dart#L43) request doesn't support this because `fields` is a `Map<String, String>` while the more correct type would be `List<Map<String, String>>` or `Map<String, List<String>>`. Unfortunately dart recently tagged the http client package with v1.0.0 so they most likely will not accept a fix even if we submit one.

Our options seems to be:

- Don't use `http.MultipartRequest` and manually construct the `multipart/form-data` request and its boundary signature

- "Fix" it on server-side by parsing brackets suffixed field keys and treat them as array

I'll try to investigate it further a little bit later, but for now the workaround is to send 2 requests if the body need to submit list values.

**Reply by ganigeorgiev** [reply]
@iamdineshkumar I've modified locally the `http.MultupartRequest` to add support for list field values. The "fix" is available with [Dart SDK v0.10.2](https://pub.dev/packages/pocketbase/versions/0.10.2).

_Just for info, there is actually an open issue in the [`dart-lang/http` repo #277](https://github.com/dart-lang/http/issues/277) but it is ~3 years old and I doubt that they will fix it because it will most likely cause a breaking change._

**Reply by iamdineshkumar** [reply]
I have tested v0.10.2 sdk and its sending list field data in request and things are working fine... but getting raw data after create hook is not providing request data
```go
requestData := apis.RequestData(e.HttpContext)
rawData, err := json.Marshal(requestData.Data)

```

**Reply by ganigeorgiev** [reply]
@iamdineshkumar make sure that you are using the latest PocketBase version as there were some internal changes related to `apis.RequestData` in recent versions. If updating your PocketBase version doesn't help, please provide a more complete reproducible code sample.

**Reply by iamdineshkumar** [reply]
I am using pockerbase 0.16.5 which is actually latest.
While member is created i am sending amount isPaid and paidDate as additional field and in case isPaid is true i am creating record in another collection in after create hook of member collection... 
```go
	requestData := apis.RequestData(e.HttpContext)
	rawData, err := json.Marshal(requestData.Data)
	if err != nil {
		return apis.NewBadRequestError("Failed to read the data", err)
	}
	payment := struct {
		Amount             float64 `json:"amount"`
		IsPaid             bool    `json:"isPaid"`
		PaidDate           string  `json:"paidDate"`
	}{}
	if err := json.Unmarshal(rawData, &payment); err != nil {
		return apis.NewBadRequestError("Failed to parse the submitted data", err)
	}
```

**Reply by ganigeorgiev** [reply]
I think I understand what do you mean. Currently it seems that the internal `echo.BindBody()` will not bind all `multipart/form-data` values when the destination is a map and will always return the first value - https://github.com/labstack/echo/blob/v5_alpha/bind.go#L133. I'll investigate it further later today and will try to provide a fix.

**Reply by ganigeorgiev** [reply, marked as answer]
@iamdineshkumar The `multipart/form-data` array values request data bind should be fixed in [v0.16.6](https://github.com/pocketbase/pocketbase/releases/tag/v0.16.6) (_I'll submit a PR later in echo for v4 and v5_), but note that this may not resolve your issue as in your case you don't seem to have array values and I'm not really sure what do you mean by _"but getting raw data after create hook is not providing request data"_. If after the update you are still not able to resolve the issue, please elaborate and/or provide the exact returned error.

**Reply by iamdineshkumar** [reply]
I have checked now... with 0.16.6 its working properly...
